### PR TITLE
FocusNode that is detached should report empty ancestors

### LINKED ISSUE #192783: FocusNode.ancestors returns a stale list after the node is detached
### Steps to reproduce

Run this widget test:

```dart
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('a detached FocusNode reports stale ancestors', (tester) async {
    final outer = FocusNode(debugLabel: 'outer');
    final inner = FocusNode(debugLabel: 'inner');
    addTearDown(outer.dispose);
    addTearDown(inner.dispose);

    await tester.pumpWidget(
      Focus(
        focusNode: outer,
        child: Focus(focusNode: inner, child: const SizedBox()),
      ),
    );
    expect(inner.ancestors, contains(outer)); // populates the cache

    // Remove the inner Focus widget; its node is detached.
    await tester.pumpWidget(Focus(focusNode: outer, child: const SizedBox()));

    expect(inner.parent, isNull);
    expect(inner.ancestors, isEmpty); // fails
  });
}
```

### Expected results

Once a node is detached, `parent` is null, so `ancestors` is empty.

### Actual results

`parent` is null, but `ancestors` still returns the full chain from before detachment:

```
parent=null
ancestors=[FocusNode#f506a(outer), FocusScopeNode#4a056(View Scope), _FocusTraversalGroupNode#32acc(FocusTraversalGroup), FocusScopeNode#d839c(Root Focus Scope)]
context=StatefulElement#901c8(DEFUNCT)(no widget)
```

`FocusNode._removeChild` (`packages/flutter/lib/src/widgets/focus_manager.dart`) sets `node._parent = null`, and clears the enclosing-scope cache and the ancestors' `_descendants`. It does not reset the removed node's own `_ancestors` cache, or that of the removed node's descendants. `_reparent` and `_updateManager` do reset it, but detachment goes through neither.

### Why it matters

Code that asks "is this node still inside that subtree?" with `node.ancestors.contains(x)` gets `true` for a node whose widget is gone. Checking `node.context != null` doesn't help, because a detached node keeps a reference to its defunct element.

We hit this in a lost-focus handler. When a focused button was removed by a screen swap, the handler decided the old node was still in the screen and called `requestFocus()` on it. That does nothing on a detached node, so keyboard focus was left nowhere and a screen reader announced nothing. The workaround is to walk `parent` links instead of reading `ancestors`.

A likely fix is to clear `_ancestors` for the removed node, and its descendants, in `_removeChild`.

### Code sample

The test above is the complete sample.

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.47.3, on macOS 26.6.2 25G83 darwin-arm64, locale en-US)
    Framework revision e8113bf456, Engine revision 06a2e2a110
```

</details>

### PR DESCRIPTION
**Pull request by navaronbracke**
This PR fixes a bug where `FocusNode.detach()` did not clear the inner ancestors list of the node.

Fixes https://github.com/flutter/flutter/issues/192783

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [AI contribution guidelines] and understand my responsibilities, or I am not using AI tools.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant in-code documentation (doc comments with `///`).
- [x] If this PR introduces a new feature or capability, I created and linked a website documentation issue or PR in [flutter/website] (or verified none is needed).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

If this change needs to override an active code freeze, provide a comment explaining why. The code freeze workflow can be overridden by code reviewers. See pinned issues for any active code freezes with guidance.

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[AI contribution guidelines]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#ai-contribution-guidelines
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[flutter/website]: https://github.com/flutter/website
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 1]
```diff
@@ -735,7 +736,13 @@ class FocusNode with DiagnosticableTreeMixin, ChangeNotifier {
   /// Iterates the ancestors of this node starting at the parent and iterating
   /// over successively more remote ancestors of this node, ending at the root
   /// [FocusScopeNode] ([FocusManager.rootScope]).
+  ///
+  /// Returns an empty iterable if this node has no parent.
   Iterable<FocusNode> get ancestors {
```
Comment (navaronbracke) [inline packages/flutter/lib/src/widgets/focus_manager.dart:738]: While the test verifies that `get ancestors` returns an empty Iterable upon detachment, I don't know if there's an easy way to test that `detach()` actually set the private backing field to null?

Happy to get some pointers on that. (Maybe with `as dynamic` ?)

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 2]
```diff
@@ -235,6 +235,7 @@ class FocusAttachment {
       // This node is no longer in the tree, so shouldn't send notifications anymore.
       _node._manager?._markDetached(_node);
       _node._parent?._removeChild(_node);
+      _node._ancestors = null;
```
Comment (Mairramer) [inline packages/flutter/lib/src/widgets/focus_manager.dart]: Hmm, instead of clearing the `_ancestors` cache in `FocusAttachment.detach`, what do you think about moving this invalidation into `FocusNode._removeChild`? Setting `node._ancestors = null` immediately after `node._parent = null` would ensure the cache is consistently invalidated whenever a node loses its parent.

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 1 above)
Comment (Mairramer) [reply packages/flutter/lib/src/widgets/focus_manager.dart:738]: I could be wrong, but since `_ancestors` is private to the library, using `dynamic` in a test would result in a `NoSuchMethodError`. The only way to test this memory retention is to keep a `WeakReference` to a previous ancestor and verify that its target becomes null after garbage collection.

Maybe if this [suggestion](https://github.com/flutter/flutter/pull/192811/changes#r4018462419) is valid, we won't need to worry about this after all.

Comment (Mairramer) [reply packages/flutter/lib/src/widgets/focus_manager.dart:738]: We’ll indeed get a `NoSuchMethodError` here.

```dart
The following NoSuchMethodError was thrown running a test:
Class 'FocusNode' has no instance getter '_ancestors'.
Receiver: Instance of 'FocusNode'
Tried calling: _ancestors
```

Comment (navaronbracke) [reply packages/flutter/lib/src/widgets/focus_manager.dart:738]: Yeah when I tried something similar, I got this too.

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 2 above)
Comment (navaronbracke) [reply packages/flutter/lib/src/widgets/focus_manager.dart]: I'll look into that!

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 3]
```diff
@@ -1027,6 +1027,7 @@ class FocusNode with DiagnosticableTreeMixin, ChangeNotifier {
     }
 
     node._parent = null;
+    node._ancestors = null;
```
Comment (Mairramer) [inline packages/flutter/lib/src/widgets/focus_manager.dart:1030]: Hmm, although `node._ancestors = null` correctly invalidates the cache for the node being removed, it doesn't clear the `_ancestors` cache for its descendants, which was explicitly mentioned in the original issue (#192783).

To keep this consistent with how `ancestor._descendants = null` is handled below, we should also invalidate the cache for all descendants:

```dart
node._parent = null;
node._ancestors = null;

for (final FocusNode descendant in node.descendants) {
  descendant._ancestors = null;
}
```

Comment (Mairramer) [reply packages/flutter/lib/src/widgets/focus_manager.dart:1030]: I believe we can use `_updateManager`, which already handles this:

```dart
node._updateManager(null);
```

Comment (navaronbracke) [reply packages/flutter/lib/src/widgets/focus_manager.dart]: Done

**Review by Mairramer — APPROVED** [review]
LGTM
